### PR TITLE
added traits for multi dimensional arrays

### PR DESCRIPTION
**Pull request by Blumerline**
this includes

- type trait __isMultiDimensionalArray__: test if a type is a multi dimensional array.
- type trait __dimensionCount__: returns the count of dimension an array has.

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -5423,6 +5425,58 @@ unittest
...
+    else
+    {
+        alias DT = typeof(T.init[0]);
+        enum isMultiDimensionalArray = hasLength!DT || isNarrowString!DT;
+    }
+}
+///
+unittest
+{
+    assert(isMultiDimensionalArray!(string[]) );
```
Comment (JackStouffer) [inline std/traits.d]: `static assert`

File: std/traits.d [hunk 2]
```diff
@@ -5423,6 +5425,58 @@ unittest
...
+    assert(!isMultiDimensionalArray!(int[]) );
+    assert(isMultiDimensionalArray!(int[][]) );
```
Comment (JackStouffer) [inline std/traits.d]: Nitpick: for doc readability, could you move the `asserts` that pass next to each other and the same for the ones that don't. It just makes it easier to scan docs.

**Comment by JackStouffer** [comment]
I want to hear what other people think before you do anything, but I believe that `dimensionCount` would fit better in `std.array`.

**Comment by Blumerline** [comment]
If you (the phobos people) are interested by dimensionCount, I'd like to have your mind about returning `0` when `T` is not an array instead of using a template constraint. This would allow to write nicer branches.

Otherwise I'm pretty sure that _isMultiDimensionalArray_ should be added and will be usefull but obviously my point of view is biased...

**Comment by Bolpat** [comment]
I'm rather a newbie and I'd expect `dimensionCount!T` to exist as enum for any type where non-array types simply have `dimensionCount == 0`. So, in this interpretation, `isArray` is just an abbreviation for `dimensionCount > 0` and `isMultiDimensionalArray` for `dimensionCount > 1`. But I find it very confusing that `isMultiDimensionalArray` needs *more than* 1 dimension to be true. For most mathematicians including myself "multiple" does not necessarily imply "more than one." But `isArray` does this job already. I'd call it `is2DArray` or `is2Array` since any 3+ dimensional array is a 2-dimensional array on top.
Wherever you need at least two dimensions, it's not too hard to write `dimensionCount!T > 1` and it's never confusing.

**Comment by Blumerline** [comment]
Ok for the enum.

As for the _multi_ prefix in `isMultiDimensionalArray`, it is used correctly here, linguistically speaking, _multi_ means many, so more than one, in opposition to _uni_. Anyway, without a math background it's crystal-clear...I don't hide that's I'm very surpised by your comment.

**Comment by Bolpat** [comment]
Suggestion:
```d
template dimensionCount(T)
{
    static if (is(T : R[], R))
        enum dimensionCount = 1 + dimensionCount!R;
    else
        enum dimensionCount = 0;
}
enum isMultiDimensionalArray(T) = dimensionCount!T > 1;
```

**Comment by lablanu** [comment]
Does this work with ranges?

**Comment by Blumerline** [comment]
@lablanu : no, But there is no ambiguities in the symbol name.

**Comment by JackStouffer** [comment]
> This issue was closed.

Well, that's a shame.

**Pull request closed without merging**